### PR TITLE
NTP protocol support

### PR DESCRIPTION
**Pull request by egecetin**

### CODE REVIEW

File: 3rdParty/EndianPortable/include/EndianPortable.h
Comment (clementperon) [inline 3rdParty/EndianPortable/include/EndianPortable.h:70]: IMO PCAPPP_MINGW_ENV is a bit ambigous.

```
#ifndef htonll
  #define htonll(x) ((1==htonl(1)) ? (x) : (((uint64_t)htonl((x) & 0xFFFFFFFFUL)) << 32) | htonl((uint32_t)((x) >> 32)))
#endif // htonll
#ifndef ntohll
  #define ntohll(x) ((1==ntohl(1)) ? (x) : (((uint64_t)ntohl((x) & 0xFFFFFFFFUL)) << 32) | ntohl((uint32_t)((x) >> 32)))
#endif // ntohll
```

And also this should be defined out of BYTE_ORDER == LITTLE_ENDIAN no ?

Comment (egecetin) [reply 3rdParty/EndianPortable/include/EndianPortable.h]: Updated, thanks

File: Packet++/header/NtpLayer.h
Comment (seladb) [inline Packet++/header/NtpLayer.h:425]: I'm thinking that since these methods are not unique to the NTP protocol, maybe we can move them to [`GeneralUtils.h`](https://github.com/seladb/PcapPlusPlus/blob/master/Common%2B%2B/header/GeneralUtils.h) and then provide methods like `getTransmitTimestampAsString()`, `getReceiveTimestampAsString()`, etc. WDYT?

File: Tests/Packet++Test/Tests/NtpTests.cpp
Comment (seladb) [inline Tests/Packet++Test/Tests/NtpTests.cpp:31]: If we move these methods to `GeneralUtils.h`  we can probably create a new test file `Packet++Test/Tests/GeneralUtilsTests.cpp` and move this test case there

File: Packet++/src/NtpLayer.cpp
Comment (seladb) [inline Packet++/src/NtpLayer.cpp:512]: maybe to keep this string short we can do something similar to Wireshark:
`NTP Layer v4, mode: client`

File: Tests/Packet++Test/Tests/NtpTests.cpp
Comment (seladb) [inline Tests/Packet++Test/Tests/NtpTests.cpp:96]: we don't have a test for `ntpLayer->toString()`

File: Packet++/src/NtpLayer.cpp
Comment (seladb) [inline Packet++/src/NtpLayer.cpp:503]: a shorter way could be:
```cpp
return data && dataSize >= sizeof(ntp_header);
```

File: 3rdParty/EndianPortable/include/EndianPortable.h
Comment (seladb) [inline 3rdParty/EndianPortable/include/EndianPortable.h:215]: I didn't see that we use it anywhere in `NtpLayer`, why add it in this PR?

File: Tests/Packet++Test/Tests/NtpTests.cpp
Comment (seladb) [inline Tests/Packet++Test/Tests/NtpTests.cpp:62]: If you want to find NTP packet that have more diverse information you can see here: https://weberblog.net/wp-content/uploads/2018/08/NTP-IPv6-with-Auth.zip

File: Packet++/header/NtpLayer.h
Comment (seladb) [inline Packet++/header/NtpLayer.h:81]: This is nice!
However, what I meant in https://github.com/seladb/PcapPlusPlus/pull/788#discussion_r791472635 is that it'd be good to document each field in each struct, like we do here:
https://github.com/seladb/PcapPlusPlus/blob/master/Packet++/header/ArpLayer.h#L24-L41

That way Doxygen will create an documentation per field like this:
https://pcapplusplus.github.io/api-docs/structpcpp_1_1arphdr.html

File: Packet++/src/NtpLayer.cpp
Comment (seladb) [reply Packet++/src/NtpLayer.cpp:113]: You didn't change anything in the code, can you please elaborate one what this calculation does?

File: Packet++/src/NtpLayer.cpp
Comment (seladb) [reply Packet++/src/NtpLayer.cpp]: Right now it returns an empty string:

```cpp
std::string NtpLayer::toString() const
{
    return std::string();
    //std::stringstream ss;

    //ss << "NTP Layer v" << (int)getVersion() << ", Mode: " << getModeString();

    //return ss.str();
}
```

I guess we should fix it?

**Comment by seladb** [comment]
@egecetin as a general note I wanted to thank you for adding NTP support, it's a great contribution to this project! 🙏 🙏

**Comment by egecetin** [comment]
@seladb During the latest commit somehow `dhcpv6creation` and test with related pcap write failed for windows. I will temporarily disable them until the update `NtpLayer` comments/reviews. Please do not merge until fix the issue.

**Comment by seladb** [comment]
> @seladb During the latest commit somehow `dhcpv6creation` and test with related pcap write failed (actually hangs during test run, possibly segv etc.) for windows. I will temporarily disable them until the update `NtpLayer` comments/reviews. Please do not merge until fix the issue.

Sure @egecetin , sounds good 👍

**Comment by seladb** [comment]
@egecetin I see you're still working on some changes. Let me know when the PR is ready for re-review and I'll take another look

**Comment by egecetin** [comment]
@seladb I think I'm done for now. Can you check again for anything i missed?

File: Packet++/header/NtpLayer.h
Comment (seladb) [inline Packet++/header/NtpLayer.h:21]: We can move these consts to `NtpLayet.cpp`, they're not used in the header file

File: Packet++/header/NtpLayer.h
Comment (seladb) [inline Packet++/header/NtpLayer.h:12]: I think we can move all of these includes to `NtpLayer.cpp` as they're not used in the header file

File: Packet++/header/NtpLayer.h
Comment (seladb) [inline Packet++/header/NtpLayer.h:5]: Is this header used somewhere?

File: Packet++/header/NtpLayer.h
Comment (seladb) [inline Packet++/header/NtpLayer.h:640]: Let's add a documentation for these methods also

**Comment by egecetin** [comment]
@seladb Let's continue from here to follow up easily. 

Updated regarding your reviews. Except this,

> We don't have tests for all the `set...InSecs` methods (`setReferenceTimestampInSecs`, `setReceiveTimestampInSecs`, `setRootDelayInSecs`, `setRootDispersionInSecs`). Maybe the easiest way would be to add to `NtpCraftingTests` crafting of a NTPv3 packet which uses these methods 🤔

_Originally posted by @seladb in https://github.com/seladb/PcapPlusPlus/pull/788#discussion_r795163539_

And my comment,

> 
> @seladb set....InSecs methods have some fundamental problem. Since they are using floating point, same as get...InSecs functions their result may or may not equal to provided value so the tests can fail. Because of this i lower the precision of this to 4 digits too. Because it can change sometimes from platform to platform too.
> 
> Because of the same problem NTP timestamp format interprets integer and fraction parts separately. I can replace these double function calls with structs,
> 
> struct X
> {
> uint32_t integer,
> float fraction
> }
> but this means they will equal to plain set....() methods because it is actually NTP Timestamp format.
> 
> Do you have any suggestions for to test these methods?

**Comment by egecetin** [comment]
@seladb And thanks for your reviews🙏

File: Packet++/header/NtpLayer.h
Comment (seladb) [reply Packet++/header/NtpLayer.h]: Now that `ntp_header` is private we need to remove this part: `(will be casted to @ref ntp_header)` because:
1. The user doesn't know what is `ntp_header` (this struct isn't exposed to him)
2. Doxygen gives an error because this struct is private

File: Packet++/header/NtpLayer.h
Comment (seladb) [inline Packet++/header/NtpLayer.h:178]: These enum values are public and not documented, hence Doxygen can't document them:
image

File: Packet++/header/NtpLayer.h
Comment (seladb) [inline Packet++/header/NtpLayer.h:27]: nit: this asterisk (`*`) is redundant:
image

File: Packet++/header/NtpLayer.h
Comment (seladb) [inline Packet++/header/NtpLayer.h:73]: nit: this asterisk (`*`) is also redundant:
image

File: Packet++/header/NtpLayer.h
Comment (seladb) [inline Packet++/header/NtpLayer.h:332]: We don't have tests for this method

File: Packet++/header/NtpLayer.h
Comment (seladb) [reply Packet++/header/NtpLayer.h:431]: > @seladb Let's continue from here to follow up easily. Searching from reviews a bit confusing
> 
> Updated regarding your reviews. Except this,
> 
> > We don't have tests for all the `set...InSecs` methods (`setReferenceTimestampInSecs`, `setReceiveTimestampInSecs`, `setRootDelayInSecs`, `setRootDispersionInSecs`). Maybe the easiest way would be to add to `NtpCraftingTests` crafting of a NTPv3 packet which uses these methods 🤔
> 
> _Originally posted by @seladb in [#788 (comment)](https://github.com/seladb/PcapPlusPlus/pull/788#discussion_r795163539)_
> 
> And my comment,
> 
> > @seladb set....InSecs methods have some fundamental problem. Since they are using floating point, same as get...InSecs functions their result may or may not equal to provided value so the tests can fail. Because of this i lower the precision of this to 4 digits too. Because it can change sometimes from platform to platform too.
> > Because of the same problem NTP timestamp format interprets integer and fraction parts separately. I can replace these double function calls with structs,
> > struct X
> > {
> > uint32_t integer,
> > float fraction
> > }
> > but this means they will equal to plain set....() methods because it is actually NTP Timestamp format.
> > Do you have any suggestions for to test these methods?

If I understand correctly the value may change from platform to platform because the accuracy of `double` varies between platforms, is that correct?
In that case, can we use an input value that is low enough and will be supported on all platforms?

**Review by seladb — COMMENTED** [review]
@egecetin I marked all resolved comments as "resolved". So all the ones that are still open are left

**Comment by seladb** [comment]
For your convenience, here are the comments currently open:
- https://github.com/seladb/PcapPlusPlus/pull/788#discussion_r795161351
- https://github.com/seladb/PcapPlusPlus/pull/788#discussion_r796291181
- https://github.com/seladb/PcapPlusPlus/pull/788#discussion_r795162934
- https://github.com/seladb/PcapPlusPlus/pull/788#discussion_r796297936
- https://github.com/seladb/PcapPlusPlus/pull/788#discussion_r795163919
- https://github.com/seladb/PcapPlusPlus/pull/788#discussion_r795164057
- https://github.com/seladb/PcapPlusPlus/pull/788#discussion_r796292108
- https://github.com/seladb/PcapPlusPlus/pull/788#discussion_r796292770
- https://github.com/seladb/PcapPlusPlus/pull/788#discussion_r796293014
- https://github.com/seladb/PcapPlusPlus/pull/788#discussion_r796294877

**Comment by egecetin** [comment]
I was trying to understand why you marking `EndianPortable.h` over and over. Then i noticed some of my comments are waiting at pending 🙄.

File: 3rdParty/EndianPortable/include/EndianPortable.h
Comment (seladb) [reply 3rdParty/EndianPortable/include/EndianPortable.h:215]: ok got it, thanks!

File: Packet++/header/NtpLayer.h
Comment (seladb) [reply Packet++/header/NtpLayer.h:5]: In that case, I think we can move `#include "IPv4Layer.h"` to `NtpLayer.cpp` 🤔

File: Packet++/header/NtpLayer.h
Comment (seladb) [reply Packet++/header/NtpLayer.h]: The argument of the `setReferenceIdentifier()` method is `uint32_t`. If we think it should be both IPv4 or `ClockSource`, why not create 2 overloads of this method - one that takes an `IPv4Address` as an argument and the other that takes `ClockSource` as an argument?

File: Packet++/header/NtpLayer.h
Comment (seladb) [reply Packet++/header/NtpLayer.h]: We still need to fix this (doxygen give an error)

**Comment by seladb** [comment]
@egecetin thank you for all of your efforts!!

Here are the comments that remain open:
- https://github.com/seladb/PcapPlusPlus/pull/788#discussion_r801323417
- https://github.com/seladb/PcapPlusPlus/pull/788#discussion_r801315931
- https://github.com/seladb/PcapPlusPlus/pull/788#discussion_r795163919
- https://github.com/seladb/PcapPlusPlus/pull/788#discussion_r801314527

Once we fix them I think we can merge the PR

File: Packet++/header/NtpLayer.h
Comment (egecetin) [reply Packet++/header/NtpLayer.h]: I already moved them. Probably you checked from outdated version

Comment (seladb) [reply Packet++/header/NtpLayer.h]: I still see:

```cpp
#include "GeneralUtils.h"
#include "SystemUtils.h"
```

In the header file, but I think they're only used in `NtpLayer.cpp` 🤔

Comment (egecetin) [reply Packet++/header/NtpLayer.h]: Understand now. I thought you comment for math.h

File: Packet++/header/NtpLayer.h
Comment (egecetin) [reply Packet++/header/NtpLayer.h]: Replaced with `#include "IpAddress.h"` because of `void setReferenceIdentifier(IPv4Address val);` since we only need `IPv4Address`

**Review by seladb — APPROVED** [review]

**Comment by seladb** [comment]
thank you so much for all of your hard work @egecetin ! I really appreciate your efforts and I'm sorry for the time it took to review and merge this PR

**Pull request merged**
